compare_model_stub_fx API implementation

DeepDriving opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48951 compare_model_stub_fx API implementation**

compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

Commit:
compare_model_stub_fx API implementation

compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

hx89 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
 
     return weight_dict
+
+def _get_logger_dict_helper(mod, target_dict, prefix=""):
```
Can we import this function from eager mode numeric suite and use it here? Would be good to reuse the existing function if it can be shared by eager mode and fx.

DeepDriving replied:
Done

hx89 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
+        _get_logger_dict_helper(child, target_dict, module_prefix)
+
+
+def get_logger_dict(mod, prefix=""):
```
ditto

DeepDriving replied:
Done

hx89 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
+    return target_dict
+
+
+class Logger(nn.Module):
```
ditto

DeepDriving replied:
Done

hx89 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
+        pass
+
+
+class ShadowLogger(Logger):
```
ditto

DeepDriving replied:
Done. Taken out of the code

hx89 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
+        self.stats["float"].append(y.detach())
+
+
+def _convert_tuple_to_list(t):
```
ditto

DeepDriving replied:
https://www.internalfb.com/diff/D25507933

hx89 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
+    return list(_convert_tuple_to_list(x) for x in t) if type(t) is tuple else t
+
+
+def _dequantize_tensor_list(t):
```
ditto

DeepDriving replied:
Done.  Taken out of the code.

DeepDriving commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
+        "quantization_api._numeric_suite.compare_model_stub_fx"
+    )
+
+    float_model = remove_qconfig_observer_fx(float_model)
```
Actually, for q_model,it is already clean. i.e. no observer is attached.  I tried to call remove_qconfig_observer on q_model and it does not have any effect.

jerryzh168 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +137,204 @@ def compare_weights_fx(float_dict, quantized_dict):
+        return output
+
+
+def prepare_model_with_stubs_fx(float_module, q_module, module_swap_list, Logger):
```
this is pretty similar to eager mode, can you just call eager mode api on a fx graphmode quantized model for this?

DeepDriving replied:
Yes, they are the same in eager mode.  My inclination is to keep fx version of prepare_model_xxx api and compare_model_xxx separate from eager mode.  Because for another fx api: compare_model_output_fx , prepare_model_output is different from eager mode and we have the API in fx.  So to make it consistent, I was thinking to  have it reside in fx.

hx89 replied:
Since prepare_model_with_stubs_fx() will be a top level API users will use, maybe good to keep "_fx" suffix to be consistent with other fx numeric suite APIs? Also it would be more flexible if we want to make fx specific changes to this function in the future.

jerryzh168 replied:
oh I mean you can call eager mode prepare_model_with_stubs here without copying the code. I think the actual fx implementation will not be able to reuse this code any ways

DeepDriving replied:
Oh I see. We can make it as a wrapper for prepare_model_with_stubs for now, and make specific changes to this function in the future if requires so. How does that sound?

jerryzh168 replied:
yes, that sounds good, I think we'll need to re-implement in the future when we want to enable some features that can only be supported by fx

raghuramank100 replied:
Makes sense, we could still reuse the eager code for modules and add additional fx specific code in the body of this function

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -142,3 +144,90 @@ def compare_and_validate_results(float_model, q_model):
+                float_model, q_model, module_swap_list, data
+            )
+            self.assertEqual(len(ob_dict), 1)
+            for k, v in ob_dict.items():
```
Would be good to add the check ob_dict.keys() are as expected.

DeepDriving replied:
Done

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -142,3 +144,90 @@ def compare_and_validate_results(float_model, q_model):
+                float_model, q_model, module_swap_list, data
+            )
+            self.assertEqual(len(ob_dict), 1)
+            for k, v in ob_dict.items():
```
Would be good to add the check ob_dict.keys() are as expected.

DeepDriving replied:
Done

raghuramank100 commented on torch/quantization/_numeric_suite_fx.py:
```diff
@@ -137,3 +138,88 @@ def compare_weights_fx(float_dict, quantized_dict):
                 )
 
     return weight_dict
+
```
Question on compare_weights_fx:  This code is identical to the eager mode implementation https://github.com/pytorch/pytorch/blob/master/torch/quantization/_numeric_suite.py#L49. For now, can we just have compare_weights_fx call compare_weights?

DeepDriving replied:
Yes, they are the same .  I am calling compare_weights inside compare_weights_fx call for now.

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -142,3 +144,96 @@ def compare_and_validate_results(float_model, q_model):
+    def test_compare_model_stub_conv_static_fx(self):
+        r"""Compare the output of static quantized conv layer and its float shadow module"""
+
+        def calibrate(model, calib_data):
```
Seems this function is the same for all tests, could you move it out so that you don't need to define it in each test?

hx89 replied:
Or use test_only_eval_fn as in your other PR?

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -142,3 +144,96 @@ def compare_and_validate_results(float_model, q_model):
+                float_model, q_model, module_swap_list, data
+            )
+            expected_ob_dict_keys = {"conv.stats"}
+            print(ob_dict.keys())
```
Remove?

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -142,3 +144,96 @@ def compare_and_validate_results(float_model, q_model):
 
         compare_and_validate_results(backup_prepared_model, q_model)
+
+    @skipIfNoFBGEMM
```
Use override_qengines to be consistent?

DeepDriving replied:
I changed to override_qeninges, calibrate to test_only_eval for all API in another PR (#49266). I can update it in this PR as well

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -142,3 +144,96 @@ def compare_and_validate_results(float_model, q_model):
+                self.img_data_2d[0][0],
+            )
+
+    @skipIfNoFBGEMM
```
Use override_qengines to be consistent?

DeepDriving replied:
I changed to override_qeninges for all APIs in other PR . Also calibrate to test_only_eval  (#49266)

jerryzh168 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -142,3 +144,96 @@ def compare_and_validate_results(float_model, q_model):
+            calibrate(prepared_model, self.img_data_2d)
+            q_model = convert_fx(prepared_model)
+
+            module_swap_list = [nn.Conv2d, nn.intrinsic.modules.fused.ConvReLU2d]
```
nn.intrinsic.modules.fused.ConvReLU2d --> nn.intrinsic.ConvReLU2d?

jerryzh168 replied:
or you can add an
import torch.nn.intrinsic as nni
and use nni here.

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]

Commit:
Update on "compare_model_stub_fx API implementation"


compare_model_stub_fx API implementation

Differential Revision: [D25379000](https://our.internmc.facebook.com/intern/diff/D25379000/)

[ghstack-poisoned]
